src/vdom.ffi.mjs

```diff
@@ -333,85 +342,18 @@ function createElementNode({ prev, next, dispatch, stack }) {
-      // in the incoming children, then we remove it. We keep doing this until we
-      // find a keyed child that we should preserve and then move on with the
-      // morph as normal.
-      while (
-        prevChild &&
-        !incomingKeyedChildren.has(prevChild.getAttribute("data-lustre-key"))
-      ) {
-        const nextChild = prevChild.nextSibling;
-        el.removeChild(prevChild);
-        prevChild = nextChild;
-      }
-
-      // If there were no keyed children in the previous render then we can just
-      // insert the incoming child at the current position (and diff against whatever
-      // is already there).
-      if (keyedChildren.size === 0) {
-        stack.unshift({ prev: prevChild, next: child, parent: el });
-        prevChild = prevChild?.nextSibling;
-        continue;
-      }
-
-      // If we come across a child that has the same key as something else this
-      // render then we can't do any meaningful morphing. We throw a warning and
-      // fall back to inserting the new node.
-      if (seenKeys.has(child.key)) {
-        console.warn(`Duplicate key found in Lustre vnode: ${child.key}`);
-        stack.unshift({ prev: null, next: child, parent: el });
-        continue;
-      }
-
-      // The `seenKeys` set is how we track duplicate keys.
-      seenKeys.add(child.key);
-      // If we make it this far then there is potentially a keyed child we can
-      // reuse from the previous render.
-      const keyedChild = keyedChildren.get(child.key);
-
-      // This case is hit when the previous render had *no* children at all. In
-      // that case we can just insert the incoming child.
-      if (!keyedChild && !prevChild) {
-        stack.unshift({ prev: null, next: child, parent: el });
-        continue;
-      }
-
-      // This is a new keyed child that wasn't in the previous render. Because we
-      // can't guarantee things won't get moved around we insert a placeholder node
-      // that preserves the position of the incoming child.
-      if (!keyedChild && prevChild !== null) {
-        const placeholder = document.createTextNode("");
-        el.insertBefore(placeholder, prevChild);
-        stack.unshift({ prev: placeholder, next: child, parent: el });
-        continue;
-      }
-
-      // This is the same as the unkeyed case: we don't have to do any special
-      // handling, just diff against the previous child and move on.
-      if (!keyedChild || keyedChild === prevChild) {
-        stack.unshift({ prev: prevChild, next: child, parent: el });
+    iterateElement(child, (currElement) => {
+      // A keyed morph has more complex logic to handle: we need to be grabbing
+      // same-key nodes from the previous render and moving them to the correct
+      // position in the DOM.
+      if (child.key !== undefined && seenKeys !== null) {
+        prevChild = diffKeyedChild(prevChild, currElement, el, stack, incomingKeyedChildren, keyedChildren, seenKeys);
```

Comment:
I moved this to share logic, but now the logic can be shared by using `iterateElement`
I didn't move it back just in case it makes the whole thing harder to understand, but it is only used here so it is totally reasonable to shift it back here